File: src/effs/udp_recv_from.c
// UDP
// ===
//! use ./sys.c

IoFall udp_recv_from(IoHand socket, uint32_t max, char* buf,
  uint32_t* len, char* host, uint32_t* port) {
  int fd = io_sys_read(socket, IO_UDPS);
  if (fd < 0) {
    return io_sys_fall(EBADF);
  }
  struct sockaddr_in at;
  socklen_t alen = sizeof(at);
  io_sync();
  ssize_t n = recvfrom(fd, buf, max, 0, (struct sockaddr*)&at, &alen);
  if (n < 0) {
    return io_sys_fall((uint32_t)errno);
  }
  *len = (uint32_t)n;
  inet_ntop(AF_INET, &at.sin_addr, host, 16);
  *port = ntohs(at.sin_port);
  return io_sys_done();
}

Term udp_recv_from_run(Env e, Term* f) {
  IoHand socket = io_hand_c(e, f[0]);
  uint32_t max = (uint32_t)f[1];
  char* buf = io_mem(malloc((uint64_t)max + 1));
  uint32_t len = 0;
  char host[16] = "";
  uint32_t port = 0;
  IoFall q = udp_recv_from(socket, max, buf, &len, host, &port);
  Term r = 0;
  if (q.code != 0) {
    r = io_fail(e, q);
  } else {
    Term stamp = io_tup(e, io_str(e, host, strlen(host)),
      io_tup(e, (uint64_t)port, io_str(e, buf, len)));
    r = io_done(e, stamp);
  }
  free(buf);
  return io_tup(e, io_hand(e, CID_SOCKET, socket), r);
}

static void __attribute__((constructor)) udp_recv_from_use(void) {
  io_eff(FID_UDP_RECV_FROM, CID_UDP_RECV_FROM, udp_recv_from_run);
}

File: bend2/effs/udp_recv_from.js
// UDP
// ===

function udp_recv_from(socket, max) {
  const sys = io_sys();
  const fd = socket;
  const b = new Uint8Array(Math.max(Number(max), 1));
  const peer = new Uint8Array(16);
  const len = new Uint32Array([16]);
  const peer_at = sys.ptr(peer);
  const peer_ln = sys.ptr(len);
  const got = sys.recvfrom(fd, sys.ptr(b), Number(max), 0, peer_at, peer_ln);
  const n = Number(got);
  if (n < 0) {
    return io_tup(socket, io_fail(sys.errno()));
  }
  const host = peer[4] + "." + peer[5] + "." + peer[6] + "." + peer[7];
  const port = (peer[2] << 8) | peer[3];
  return io_tup(socket, io_done(io_tup(host, port, io_text(b, n))));
}

function udp_recv_from_need() {
  return { read: true };
}
